engine/runtime-compiler/src/main/java/org/enso/compiler/pass/MiniIRPass.java

```diff
@@ -0,0 +1,58 @@
+package org.enso.compiler.pass;
+
+import java.util.Set;
+import org.enso.compiler.core.IR;
+
+/**
+ * Mini IR pass operates on a single IR element at a time. The {@link org.enso.compiler.Compiler}
+ * traverses the whole IR tree bottom-up from leaves to the root. For each IR element:
+ *
+ * <ol>
+ *   <li>The {@link #prepare(IR)} method is called to prepare the pass for the current IR element.
+ *       This method is called when the {@link org.enso.compiler.Compiler} traverses the element
+ *       from top to bottom. This is useful for mini passes that need to build some information
+ *       about the current IR element before transforming it.
+ *   <li>The {@link #transformIr(IR)} method is called to transform the current IR element. This
+ *       method is called when the {@link org.enso.compiler.Compiler} traverses the element from
+ *       bottom to top. All the children of the current IR element are already transformed when this
+ *       method is called.
+ * </ol>
+ *
+ * <p>Inspired by: <a href="https://dl.acm.org/doi/10.1145/3140587.3062346">Miniphases: compilation
+ * using modular and efficient tree transformations</a>. PDF available at <a
+ * href="https://infoscience.epfl.ch/server/api/core/bitstreams/8ab72c0a-8aa6-4dee-a704-3504938dc316/content">infoscience.epfl.ch</a>
+ */
+public abstract class MiniIRPass {
+
+  /** Set of mini passes that should run before this mini pass. */
+  public Set<MiniIRPass> runsAfter() {
+    return Set.of();
+  }
+
+  /**
+   * Prepare the pass for the provided IR element. This method is called when the {@link
+   * org.enso.compiler.Compiler} traverses the IR element from top to bottom.
+   *
+   * @param ir IR element to be prepared for transformation.
+   */
+  public void prepare(IR ir) {
```

Comment:
Copied from the paper, right? I have to admit I didn't get how they wanted this to work reliably. As discussed last week, I believe we should have:
```java
public default MiniIRPass prepare(IR ir) {
  return this;
}
```
with following semantic:
- return `null` if you are no longer interested in the subtree of provided `IR`
- return `this` if you want to process the `IR` subtree (that's the default)
- create a copy of itself with additional info collected from `IR` and let the copy be used while processing the subtree

First of all we have to _find a use-case_ we need this API for, however. Meanwhile I suggest to remove the `prepare` method from the interface.